proto/p4/p4types.proto

```diff
@@ -0,0 +1,207 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package p4;
+
+// P4 type specs ---------------------------------------------------------------
+
+// From the P4_16 spec:
+/* |--------------------|--------------------------------------------|
+ * |                    | Container type                             |
+ * | Element type       |-----------|--------------|-----------------|
+ * |                    | header    | header_union | struct or tuple |
+ * |--------------------|-----------|--------------|-----------------|
+ * | bit<W>             | allowed   | error        | allowed         |
+ * | int<W>             | allowed   | error        | allowed         |
+ * | varbit<W>          | allowed   | error        | allowed         |
+ * | int                | error     | error        | error           |
+ * | void               | error     | error        | error           |
+ * | error              | error     | error        | allowed         |
+ * | match_kind         | error     | error        | error           |
+ * | bool               | error     | error        | allowed         |
+ * | enum               | error     | error        | allowed         |
+ * | header             | error     | allowed      | allowed         |
+ * | header stack       | error     | error        | allowed         |
+ * | header_union       | error     | error        | allowed         |
+ * | struct             | error     | error        | allowed         |
+ * | tuple              | error     | error        | allowed         |
+ * |--------------------|-----------|--------------|-----------------|
+ */
+
+// These P4 types (struct, header_type, header_union and enum) are guaranteed to
+// be "named" (e.g. you cannot use an anonymous struct to declare a variable,
+// like in C). Instead of duplicating the type spec for these every time the
+// type is used, we include the type spec once in this P4TypeInfo message and
+// refer to the types by name in the P4DataTypeSpec message. We also support
+// annotations for these type specs which can be useful, e.g. to identify
+// well-known headers (such as ipv4).
+message P4TypeInfo {
+  map<string, P4StructTypeSpec> structs = 1;
+  map<string, P4HeaderTypeSpec> headers = 2;
+  map<string, P4HeaderUnionTypeSpec> header_unions = 3;
+  map<string, P4EnumTypeSpec> enums = 4;
+  P4ErrorTypeSpec error = 5;
+}
+
+// Describes a P4_16 type.
+message P4DataTypeSpec {
+  oneof type_spec {
+    P4BitstringLikeTypeSpec bitstring = 1;
+    P4BoolType bool = 2;
+    // From the P4_16 spec: "A tuple is similar to a struct, in that it holds
```

Comment:
Nit: I'd prefer to see this comment right before message P4TupleSpec

Reply:
done